pkg/ring/ring_test.go

```diff
@@ -609,6 +610,356 @@ func TestRing_ShuffleShard_Consistency(t *testing.T) {
+				{what: add, instanceID: "instance-3", instanceDesc: generateRingInstanceWithInfo("instance-3", "zone-c", []uint32{userToken(userID, 2) + 1}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-4", instanceDesc: generateRingInstanceWithInfo("instance-4", "zone-a", []uint32{userToken(userID, 0) + 3}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-5", instanceDesc: generateRingInstanceWithInfo("instance-5", "zone-b", []uint32{userToken(userID, 1) + 3}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-6", instanceDesc: generateRingInstanceWithInfo("instance-6", "zone-c", []uint32{userToken(userID, 2) + 3}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-7", instanceDesc: generateRingInstanceWithInfo("instance-7", "zone-a", []uint32{userToken(userID, 0) + 2}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-8", instanceDesc: generateRingInstanceWithInfo("instance-8", "zone-b", []uint32{userToken(userID, 1) + 2}, now.Add(-2*lookbackPeriod))},
+				{what: add, instanceID: "instance-9", instanceDesc: generateRingInstanceWithInfo("instance-9", "zone-c", []uint32{userToken(userID, 2) + 2}, now.Add(-2*lookbackPeriod))},
+				{what: test, shardSize: 3, expected: []string{"instance-1", "instance-2", "instance-3"}},
+				{what: test, shardSize: 6, expected: []string{"instance-1", "instance-2", "instance-3", "instance-7", "instance-8", "instance-9"}},
+			},
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Initialise the ring.
+			ringDesc := &Desc{Ingesters: map[string]IngesterDesc{}}
+			ring := Ring{
+				cfg: Config{
+					HeartbeatTimeout:     time.Hour,
+					ZoneAwarenessEnabled: true,
+				},
+				ringDesc:         ringDesc,
+				ringTokens:       ringDesc.getTokens(),
+				ringTokensByZone: ringDesc.getTokensByZone(),
+				ringZones:        getZones(ringDesc.getTokensByZone()),
+				strategy:         &DefaultReplicationStrategy{},
+			}
+
+			// Replay the events on the timeline.
+			for _, event := range testData.timeline {
+				switch event.what {
+				case add:
+					ringDesc.Ingesters[event.instanceID] = event.instanceDesc
+
+					ring.ringTokens = ringDesc.getTokens()
+					ring.ringTokensByZone = ringDesc.getTokensByZone()
+					ring.ringZones = getZones(ringDesc.getTokensByZone())
+				case remove:
+					delete(ringDesc.Ingesters, event.instanceID)
+
+					ring.ringTokens = ringDesc.getTokens()
+					ring.ringTokensByZone = ringDesc.getTokensByZone()
+					ring.ringZones = getZones(ringDesc.getTokensByZone())
+				case test:
+					rs, err := ring.ShuffleShardWithLookback(userID, event.shardSize, lookbackPeriod).GetAll(Read)
+					require.NoError(t, err)
+					assert.ElementsMatch(t, event.expected, rs.GetAddresses())
+				}
+			}
+		})
+	}
+}
+
+func TestRing_ShuffleShardWithLookback_CorrectnessWithFuzzy(t *testing.T) {
```

Comment:
Ignore this test for now because it's not correct. I should mock `time.Now()` in `ShuffleShardWithLookback()`. I'm working on a fix.

Reply:
After some investigation I've found out that, when the shard size is increased (eg. from 3 to 6) and there are multiple zones, the "consistency" property is not guaranteed (you can end up with all new instances in the subring) and this breaks this test.

_I've pushed a version of this test that breaks because of this. I will fix the issue in a separate PR, to keep the diff easier to review._

Reply:
With the fix done in #3299 this test passes.